### DISCUSSION TITLE
Using PocketBase with client side authentication

### DISCUSSION THREAD

**Discussion by pchasco** [pocketbase/pocketbase#1104] [Q&A]
In my application, I would prefer to use native authentication mechanisms, rather than use PocketBase. What I would like to see is if we can authenticate using whatever means in our app (Google Sign In, Apple Auth, etc), and simply pass a signed ID token to the backend service. This is how Google specifies applications should be interacting with third-party backend services (https://developers.google.com/identity/sign-in/web/backend-auth), and I think this is a better approach for native applications rather than using something like flutter_web_auth.

In another post, you outlined a possible workaround to enable "auth via token." (https://github.com/pocketbase/pocketbase/discussions/725) I am somewhat unclear about how this accomplishes that functionality.

Being able to pass the signed id token from the OpenID provider to PocketBase would simplify integration for native apps, and provide a better native experience. This would require implementing validation of the token for various providers, but token validation is fairly straightforward, and can be handled automatically by the issuer of the token.

**Comment by ganigeorgiev** [comment]
I don't think we can implement this because not all providers support user id tokens. There are plans to simplify the OAuth2 integration per #55.

If you need custom auth mechanisms you can create your own middleware and authenticate the app user from there:

```go
package main

import (
    "log"

    "github.com/labstack/echo/v5"
    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
    "github.com/pocketbase/pocketbase/core"
)

func customAuth(app core.App) echo.MiddlewareFunc {
    return func(next echo.HandlerFunc) echo.HandlerFunc {
        return func(c echo.Context) error {
            val := c.Request().Header.Get("SOME-SPECIAL-HEADER")

            // ...verify that val could be trusted...

            // find and authenticate the app user
            record, err := app.Dao().FindAuthRecordByEmail("users", val)
            if err == nil {
                c.Set(apis.ContextAuthRecordKey, record)
            }

            return next(c)
        }
    }
}

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        e.Router.Use(customAuth(app))

        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by ganigeorgiev** [reply]
_As a side note, I'm currently experimenting with adding support for registering hooks with goja, so eventually you'll be able to do the above with JS outside of the PocketBase executable._

**Comment by pchasco** [comment]
Thanks! Of course you are correct, not all providers would be compatible with this approach. I may expand on what you’ve provided for my purpose and provide a package that can be utilized by others. My priorities are to support Google and Apple sign in.
